libkernelflinger/tpm2_security.c

```diff
@@ -815,6 +834,113 @@ EFI_STATUS tpm2_fuse_bootloader_policy(
 	return EFI_UNSUPPORTED;
 }
 
+static EFI_STATUS tpm2_fuse_bootloader(void)
+{
+	EFI_STATUS ret;
+	UINT16 config_index;
+	BYTE data[NV_INDEX_BOOTLOADER_SIZE] = {0};
+	BYTE data_read[sizeof(data)];
+	UINT16 data_read_size = sizeof(data);
+	tpm2_bootloader_t *bootloader = (tpm2_bootloader_t *)data;
+
+	config_index = NV_INDEX_BOOTLOADER - config_table[0].nv_index;
+	ret = tpm2_create_nvindex(NV_INDEX_BOOTLOADER, config_table[config_index].attribute, sizeof(data));
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Failed to create bootloader NV index");
+		return ret;
+	}
+
+	bootloader->struct_ver = NV_INDEX_BOOTLOADER_STRUCT_VER;
+	bootloader->lock_state = UNLOCKED;  // Note: set to unlock in a device just create the NV index.
+
+	// Read the data again to verify it
+	ret = tpm2_write_nvindex(NV_INDEX_BOOTLOADER, sizeof(data), data, 0);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Write bootloader NV index failed");
+		goto out;
+	}
+
+	// Read the data again to verify it
+	ret = tpm2_read_nvindex(NV_INDEX_BOOTLOADER, &data_read_size, data_read, 0);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Read bootloader NV index back failed just after write it");
+		goto out;
+	}
+
+	if (memcmp(data, data_read, sizeof(data))) {
+		error(L"Security error! Read bootloader NV index back but verify failed!");
+		ret = EFI_SECURITY_VIOLATION;
+		goto out;
+	}
+
+	debug(L"Success create and write bootloader NV index");
+out:
+	return ret;
+}
+
+static EFI_STATUS tpm2_check_bootloader_index(void)
+{
+	EFI_STATUS ret;
+	TPM2B_NV_PUBLIC NvPublic;
+	TPM2B_NAME NvName;
+	UINT8 struct_ver;
+	UINT16 data_size = sizeof(struct_ver);
+	UINT32 *attr;
+	UINT32 *config_attr;
+
+	// Check the SEED nvindex
+	ret = Tpm2NvReadPublic(NV_INDEX_BOOTLOADER, &NvPublic, &NvName);
+	if (EFI_ERROR(ret)) {
+		if (ret != EFI_NOT_FOUND) {
+			efi_perror(ret, L"Read bootloader NV index failed");
+			return ret;
+		}
+
+		// Can't find it, try to init it now
+		ret = tpm2_fuse_bootloader();
+		if (EFI_ERROR(ret))
+			efi_perror(ret, L"Failed to fuse bootloader NV index");
+
+		return ret;
+	}
+
+	// After fuse, the TPM maybe add more attribute, so need to skip them
+	NvPublic.nvPublic.attributes.TPMA_NV_WRITTEN = 0;
+	NvPublic.nvPublic.attributes.TPMA_NV_WRITELOCKED = 0;
+	attr = (UINT32 *)&NvPublic.nvPublic.attributes;
+	config_attr = (UINT32 *)&config_table[NV_INDEX_BOOTLOADER - config_table[0].nv_index].attribute;
+	if (NvPublic.nvPublic.dataSize < sizeof(tpm2_bootloader_t) || *attr != *config_attr) {
+		// Find it, but the data is empty or wrong.
+		error(L"Find bootloader NV index, but the data is wrong, data size: %d, attributes: 0x%lx, need: 0x%lx",
```

Comment:
You could resolve the too long line issue with the following construction
```
error(L"Find bootloader NV index, but the data is wrong, "
         L"data size: %d, attributes: 0x%lx, need: 0x%lx", [...]);
```

Reply:
If we force let the max line to 80, then the code will like the following:
        if (!per.ownerAuthSet) {
                error(L"Try to lock TPM owner, success call "
                                L"Tpm2HierarchyChangeAuth, but ownerAuthSet "
                                L"is not set!");
                ret = EFI_SECURITY_VIOLATION;
                goto out;
        }

And this will cause the code is more strange and easy cause the wrong log information.
Now the screen resolution is much higher and can easy to support long line, so I think 132 is more reasonable than 80.

Reply:
Well 80 is still the rule even on new projects.